dominance indices

antagomir opened the pull request:
With @tvborm we have added dominance indices in mia. 

This is conceptually close to alpha diversity indices. 

However, we chose (for now) to implement dominance indices as their own `estimateDominance` function; this makes the distinction between dominance and diversity more clear.

In a sense, dominance indices are form of alpha diversity, and therefore they could be made available also through `estimateDiversity` function (which would then internally call estimateDominance for those indices).

Opening the PR to initiate discussion. The unit tests are on the way.

antagomir committed:
Added MovingPictures data set

antagomir committed:
gitignore augmentation

antagomir committed:
MovingPictures data resave and documentation

antagomir committed:
prevalence functions

antagomir committed:
Merge pull request #1 from microbiome/antagomir

Added MovingPictures data set and add prevalence functions.

antagomir committed:
Revert "Added MovingPictures data set"

antagomir committed:
Merge pull request #2 from microbiome/revert-1-antagomir

Revert "Added MovingPictures data set"

antagomir committed:
Prevalence functions added

antagomir committed:
gitignore fix

antagomir committed:
changed prevalent function nams

antagomir committed:
as_relative

antagomir committed:
Added first prevalence tests

antagomir committed:
Prevalence man pagess

antagomir committed:
Formatting ok

antagomir committed:
Sanitizing prevalence funcs

FelixErnst committed:
cleaned up prevalence calculation and added taxonomic rank selection

FelixErnst committed:
added link between getPrevalence and getTopTaxa

antagomir committed:
Merge remote-tracking branch 'upstream/master'

antagomir committed:
Fixed agglomerateByRank for 1-rank case

antagomir committed:
fixed spelling errors

antagomir committed:
Merge branch 'agglomerateByRank' of https://github.com/microbiome/mia

antagomir committed:
fixed the PR

antagomir committed:
Fixed aggomeration

antagomir committed:
rebase

antagomir committed:
Performance improvement

antagomir committed:
DataFrame fix for performance improvement

antagomir committed:
Merge remote-tracking branch 'upstream/master'

TuomasBorman committed:
Synced with FelixErnst/mia

Merge remote-tracking branch 'upstream/master' into tvborm

# Conflicts:
#	DESCRIPTION
#	R/agglomerate.R
#	R/getPrevalence.R
#	tests/testthat/test-5prevalence.R

TuomasBorman committed:
Added estimateDominance.R

TuomasBorman committed:
Added prevalentAbundance.R

TuomasBorman committed:
Fixed the function

antagomir committed:
Merge remote-tracking branch 'upstream/master'

antagomir committed:
Merge pull request #3 from openresearchlabs/tvborm

Tvborm

antagomir committed:
run devtools::document

antagomir committed:
Augmented biocViews

antagomir committed:
explicit import for stats and utils

antagomir committed:
added roxygen man pages for the new functions

antagomir committed:
estimateDominance is OK

antagomir committed:
fixed the name length bug

antagomir committed:
added utility scripts for pkg build

FelixErnst commented on NAMESPACE:
```diff
@@ -30,6 +31,7 @@ export(makeMicrobiomeExperimentFromphyloseq)
 export(meltAssay)
 export(mergeCols)
 export(mergeRows)
+export(prevalentAbundance)
```
This is missing the verb in the function name

antagomir replied:
Yes. Should be `getPrevalentAbundance`

TuomasBorman replied:
prevalentAbundance changed to getPrevalentAbundance

FelixErnst commented on R/estimateDominance.R:
```diff
@@ -0,0 +1,366 @@
+
+
+#---------------------------Help functions----------------------------------------------------------------
+checkIndices <- function(index){
```
please use the `.` prefix for internal functions, and snake_case if you like to keep it consistent

TuomasBorman replied:
Changed the names of internal functions. E..g checkIndices --> .check_indices and gini --> .gini

FelixErnst commented on inst/extras/build.sh:
```diff
@@ -0,0 +1,8 @@
+#~/bin/R-patched/bin/R CMD BATCH document.R
```
This is not really necessary, since the package does not need to "know" how to check itself. And the version number needs to be tracked, which is more work than necessary

antagomir replied:
Yes this was my local stuff, can be removed.

FelixErnst commented on inst/extras/document.R:
```diff
@@ -0,0 +1,31 @@
+# Package release instructions: http://r-pkgs.had.co.nz/release.html
```
See above. Not needed, is it?

antagomir replied:
Same.

FelixErnst commented on R/estimateDominance.R:
```diff
@@ -0,0 +1,366 @@
+#'
+#' This function calculates the community dominance index.
+#'
+#' This includes the \sQuote{DBP}, \sQuote{DMN}, \sQuote{Absolute},
```
This sentence is not complete. Maybe a bit on the background of these values and some references, if you want to include them, might round it out.

FelixErnst commented on R/estimateDominance.R:
```diff
@@ -0,0 +1,366 @@
+#' @details
+#' \code{estimateDominance} calculates the community dominance indices.
+#'
+#' @return Returns ME object, that includes dominance indices in coldata.
```
Nope. It should return the same class a `x`, so that it doesn't matter whether you use a `SE`, `TSE` or `ME`

TuomasBorman replied:
Changed the text after return to that what is below. I took it from the estimateDiversity, and I am pretty sure (not 100 %) that it does the job.


@return \code{x} with additional \code{\link{colData}} named \code{*name*}

FelixErnst replied:
If you want make it a sentence. Always better for readability.

FelixErnst commented on R/estimateDominance.R:
```diff
@@ -0,0 +1,366 @@
+#' }
+#'
+#' @name estimateDominance
+#' @export
```
two export tags in this block

FelixErnst commented on R/estimateDominance.R:
```diff
@@ -0,0 +1,366 @@
+#' colData(esophagus)
+#'
+#' @export
+#'
```
Just as tip for clarity:

I like to separate the man page docu from the actual function, since the export tag above is only applied to the generic mentioned below. By ending this block like this

```
#' @examples
#' ... some stuff
NULL

#' @rdname estimateDominance
#' @export
setGeneric("estimateDominance",signature = c("x"),
...
```
the association becomes more clear, at least in my mind.

antagomir replied:
yes this is good.

TuomasBorman replied:
Changed to that format

FelixErnst commented on R/estimateDominance.R:
```diff
@@ -0,0 +1,366 @@
+#' @param name A name for the column of the colData where the calculated
+#' Dominance indices should be stored in.
+#'
+#' @param BPPARAM A
```
Not used in the code

TuomasBorman replied:
Removed BPPARAM argument because it is not used in the code

TuomasBorman replied:
Added it back, because it is now used in the code

FelixErnst commented on R/estimateDominance.R:
```diff
@@ -0,0 +1,366 @@
+#' @export
+setMethod("estimateDominance", signature = c(x = "MicrobiomeExperiment"),
+          function(x, abund_values = "counts", index="all", rank=1, as_relative=TRUE, aggregate=TRUE,
+                   name = index, BPPARAM = SerialParam(), ...){
```
Regarding `BPPARAM` see above

TuomasBorman replied:
Removed BPPARAM argument because it is not used in the code

FelixErnst commented on R/estimateDominance.R:
```diff
@@ -0,0 +1,366 @@
+#' @rdname estimateDominance
+#' @export
+setMethod("estimateDominance", signature = c(x = "MicrobiomeExperiment"),
+          function(x, abund_values = "counts", index="all", rank=1, as_relative=TRUE, aggregate=TRUE,
```
the types of `abund_values` `rank` `as_relative` `aggregate` are not checked. Please apply the "fail early, fail often" maxime

by the way `rank`, if you need to pass it on to the `agglomerateByRank` function needs to be a character. Use `taxonomyRanks(x)[1L]` as the default value

TuomasBorman replied:
Added testing. taxonomyRanks(x)[1L] to default value of rank

TuomasBorman replied:
Changed rank back to numerical values 1 or 2, because it is not a phylogenetic rank. It is used when calculating DBP and DMN. It stores the value 1 or 2 which represents if the most abundant taxon is taken into account (1) or if two most abundant taxa is taken into account (2).

Because rank is used represent phylogenetic rank in many other functions, it might be good idea to consider other name for that to avoid misunderstanding. 

Now, I changed the name of 'rank' to 'ntaxa'.

FelixErnst commented on R/estimateDominance.R:
```diff
@@ -0,0 +1,366 @@
+
+    #If there is indices, but none of them are accepted, gives warning and returns null
+    if (!is.null(index) && !any(index %in% accepted)) {
+        warning(paste("None of the dominance indices are not recognized:", paste(index, sep = ",")))
```
this should be an error

TuomasBorman replied:
Changed to error

FelixErnst commented on R/estimateDominance.R:
```diff
@@ -0,0 +1,366 @@
+
+
+#---------------------------Help functions----------------------------------------------------------------
+checkIndices <- function(index){
```
This whole function is to much fluff. The user is required to know, that he/she/they are doing, this includes typing correctly. Replace the function with

```
index <- match.arg(index)
```

and be done with it.

TuomasBorman replied:
Deleted function and used the command above

FelixErnst commented on R/estimateDominance.R:
```diff
@@ -0,0 +1,366 @@
+#' @rdname estimateDominance
+#' @export
+setMethod("estimateDominance", signature = c(x = "MicrobiomeExperiment"),
+          function(x, abund_values = "counts", index="all", rank=1, as_relative=TRUE, aggregate=TRUE,
```
If you define `index` likes this

```
index = c("DBP", "DMN", "absolute", "relative", "simpson", "core_abundance", "gini")
```

in the function and the generic definition, you can use `match.arg` very easily and all the options are included in the man pages, which increases readability. See comment below

TuomasBorman replied:
I agree, that it would be beneficial to use that simple solution. Although with this, I think it is not possible to calculate multiple indices at the same time. With the .check_indices, it was. 

Is it ok, that it is possible to calculate indices only one by one?

(Well, e.g. EstimateDiversity's indices are calculated also one by one, because its function are like estimateRichness. estimateShannon...)

antagomir replied:
Hmm.. but `estimateDiversity(x)` does calculate all indices by default (see 3978bbf). Isn't this what you are after as well?

FelixErnst replied:
just do a `match.arg(index, several.ok = TRUE)`. Then you get a vector back, which can only contain values as defined in the function definition. If you then do a lapply on those you should be able to get all the values you need. The lapply can also be switched to `BiocParallel::bplapply` so that the values for the different indices are calculated in parallel

FelixErnst replied:
You can have a look at `estimateDiversity` including the latest changes, if you want to implement parallelization. However, parallelization makes only sense, if there are more then 1000 samples and mostly on non-windows systems.

antagomir replied:
i.m.o would make sense to implement it since large collections are increasingly common now

TuomasBorman replied:
Ok, thanks for clarifying. I replaced the check_indices function with this more simple solution

TuomasBorman replied:
Implemented parallelization.

FelixErnst commented on R/estimateDominance.R:
```diff
@@ -0,0 +1,366 @@
+dominance_help <- function(x, abund_values = "counts", index="all", rank=1, as_relative=TRUE,
+                           aggregate=TRUE) {
+    #Stores the absolute abundances to "otu" variable
+    otu <- assays(x)[[abund_values]]
```
Use `assay(x,abund_values)`

edit: This goes for all instances in the PR

TuomasBorman replied:
Changed all the assays() in the PR to assay()

FelixErnst commented on R/estimateDominance.R:
```diff
@@ -0,0 +1,366 @@
+    otu <- assays(x)[[abund_values]]
+
+    #if index does not have any values
+    if (is.null(index)) {
```
This should never be the case and is defined by the next line of code, which just non sensical

TuomasBorman replied:
The default value of index in .dominance_help (I haven't yet changed the name) is now "all". But there is no "all" index, so it does not really do anything. I changed the default value of index to NULL, because I think it makes more sense, because then the index is calculated based on other values that are provided by user (or from default).

FelixErnst replied:
change the default to all the indeces and remove the `all` value. Then all the indices will be calculated

FelixErnst replied:
Like this `index = c("DBP", "DMN", "absolute", "relative", "simpson", "core_abundance", "gini")`. add this in the generic or add all the values in `setMethod` and the `match.arg` call

TuomasBorman replied:
Sorry, if I did not understood correctly, but now the .dominance_help is just calculating one index at a time, so it is not possible to add all indices to .dominance_help's default argument.  It is like .run_dvrsty in estimateDiversity.

TuomasBorman replied:
All in all, estimateDominance works like estimateDiversity. All indices are as a default value of "index" in estimateDominance.

Like this 

#' @rdname estimateDominance
#' @export
setMethod("estimateDominance", signature = c(x = "MicrobiomeExperiment"),
          function(x, abund_values = "counts", index = c("DBP", "DMN", "absolute", "relative", "simpson", "core_abundance", "gini"),
                   ntaxa=1, as_relative=TRUE, aggregate=TRUE, name = index, ..., BPPARAM = SerialParam()){

TuomasBorman replied:
And as a default, you can calculate all indices, like with estimtaeDiversity.

#with...
esophagus <- estimateDominance(esophagus)
esophagus <- estimateDiversity(esophagus)
#...you can calculate all dominance and diversity indices which can be accessed with 
colData(esophagus)

TuomasBorman replied:
I implemented major part of estimateDiversity to estimateDominance, so only bigger difference right now is, how .dominance_help/.run_dvrsty work in detail even though they work same way from the point of user. 

E.g. there is now...

#Calculates dominance indices
              dominances <- BiocParallel::bplapply(X=index,
                            FUN = .dominance_help, x=x,
                            abund_values=abund_values, ntaxa=ntaxa,
                            as_relative=as_relative,
                            aggregate=aggregate,
                            BPPARAM = BPPARAM)

              #Add dominance indices to colData
              .add_dominances_values_to_colData(x, dominances, name)


...which is "directly" from estimateDiversity.

TuomasBorman replied:
So, do you think that also .dominance_help should get all indices as default? That would require some kind of loop/apply structure

FelixErnst replied:
Lets take a step back.

1. Is the .domincance_help function still necessary?
2. If yes what is the function of `.domincance_help`?
3. Why does `.domincance_help` need to know the value of `index`?
4. If it needs to know, than it needs to be able to handle indeces with length of `>= 1`.

So I think somewhere in this line of inquiry lies the solution. It might help, if you would push the current state into the branch, so it gets updated here and I can have a look at the current state.

Otherwise keep up the good work! 👍

TuomasBorman replied:
Oh, I did not realize that it is not updating here. I was updating only my own branch... sorry. Now, all changes should be here also

antagomir replied:
I just think that we should follow the identical conventions with estimateDominance and estimateDiversity wherever possible.

FelixErnst replied:
I agree. I am quite sorry, that I had to change things around myself, but the dominance values are getting computed in several different ways, so that it was very complex to explain. I hope you both don't mind the changes

TuomasBorman replied:
On my behalf, it is more than fine that you did those changes. Actually, I was thinking if I should implement that "last part" of estimateDiversity also because of the consistency, and there was no reason to do it differently in estimateDominance. So yeah, everything is fine

FelixErnst replied:
Cool 👍 

btw: the `.add_dvrsty_values_to_colData` function is a bit more complex, because it needs to handle nested `DataFrame`s. Since the dominance calculation always return a vector, that is not necessary

FelixErnst commented on R/estimateDominance.R:
```diff
@@ -0,0 +1,366 @@
+    } else if (index == "absolute") {
+        #Rank=1 by default but can be tuned
+        as_relative <- FALSE
+    } else if (index %in% c("relative")) {
```
index should always be of `length == 1L`. Therefore it should be `index == "relative"` as it is the case in the first if part

TuomasBorman replied:
Corrected

FelixErnst commented on R/estimateDominance.R:
```diff
@@ -0,0 +1,366 @@
+    return(index)
+}
+
+dominance_help <- function(x, abund_values = "counts", index="all", rank=1, as_relative=TRUE,
```
I don't get the function name. For me `.norm_rank` makes more sense.

However, since more than one step of the analysis is done in this function, this should be broken up into seperate functions

FelixErnst commented on R/estimateDominance.R:
```diff
@@ -0,0 +1,366 @@
+
+}
+
+gini <- function(x, abund_values="counts") {
```
Does not match its usage above. `abund_values` not passed on.

`assay(x,abund_values)` should be done externally and a matrix used as only input.

edit: like you did for `simpson_dominance`

TuomasBorman replied:
Changed the functionality that it is similar to .simpson_dominance / matrix is only input

FelixErnst commented on R/estimateDominance.R:
```diff
@@ -0,0 +1,366 @@
+
+
+
+gini_help <- function(x, w=rep(1, length(x))) {
```
Rename to `.do_gini` or `.calculate_gini`

TuomasBorman replied:
.gini_help to .calculate_gini

FelixErnst commented on R/estimateDominance.R:
```diff
@@ -0,0 +1,366 @@
+
+}
+
+gini <- function(x, abund_values="counts") {
```
Rename to `.get_gini`

TuomasBorman replied:
.gini to .get_gini

FelixErnst commented on R/estimateDominance.R:
```diff
@@ -0,0 +1,366 @@
+              #If the indices list is not null
+              if (!is.null(index)) {
+                  #loop through index list
+                  for (idx in index) {
```
Vectorize with `lapply`

edit: then you can also parallelize with `BiocParallel`

TuomasBorman replied:
Vectorized and parallized

FelixErnst commented on R/estimateDominance.R:
```diff
@@ -0,0 +1,366 @@
+              tab <- NULL
+
+              #If the indices list is not null
+              if (!is.null(index)) {
```
If `index` is `NULL`, something has gone wrong. The user should not get away with it. "fail early, fail often"

TuomasBorman replied:
Corrected

FelixErnst commented on R/estimateDominance.R:
```diff
@@ -0,0 +1,366 @@
+
+    #Checks if the length of names and indices match. If not, returns ME object without new indices.
+    if(length(names) != length(colnames(indicesDF))){
+        warning("Number of names and indices do not match.")
```
This should be an error

TuomasBorman replied:
Changed to error

FelixErnst commented on R/estimateDominance.R:
```diff
@@ -0,0 +1,366 @@
+}
+
+#Gets ME object and index data frame as parameters
+add_indices_to_coldata <- function(x, indicesDF, names){
```
This whole function can be replace by doing `colData(x) <- cbind(colData(x), indicesDF)`

Presort `indicesDF` if you need to, but this should not be required

TuomasBorman replied:
I think it is not working exactly same way. At least, something must change if that easier solution is used. 
In that point, indices have their own names i.e. "gini", "simpson"... In the ".add_indices_to_coldata" function, indices are stored in the object with the name that user has specified. It can be "gini", but it can also be "index1".

 colData(x) <- cbind(colData(x), indicesDF) does not take into account that. It just adds indices to the object. If user want that "gini" is "index1" it does not happen. 

I can change the functionality that way that user can not specify the name, so gini is always "gini". Or I can change it that way that the indices name, that user has specified, are stored as a name of indices earlier.

TuomasBorman replied:
Also, I tested it both ways, and I found out thay if user calculates same indices multiple times, same indices are stored in the object multiple times, if cbind() is used. 

In the add_indices_to_coldata, one index is saved only one time. Of course, if user wants it to be stored multiple times, user can give them different names. E.g. there can be "gini" and "gini2", but not "gini" and "gini".

FelixErnst replied:
Then try something like this:

```
colData(x)[,names] <- indicesDF
```

If the names exist they will get overwritten, if not they will get attached as long `length(names) == ncol(indicesDF)`

TuomasBorman replied:
I implemented .add_dvrsty_values_to_colData from esimateDiversity, because I thought that there is no need to do it differently.

FelixErnst commented on R/prevalentAbundance.R:
```diff
@@ -0,0 +1,76 @@
+                                     as_relative=as_relative, rank=NULL)
+
+              if (length(cm) == 0) {
+                  warning("With the given abundance and prevalence
```
This should be an error

TuomasBorman replied:
--> error

FelixErnst commented on R/prevalentAbundance.R:
```diff
@@ -0,0 +1,76 @@
+#' @rdname getPrevalence
+#' @export
+setMethod("prevalentAbundance", signature = c(x = "MicrobiomeExperiment"),
+          function(x, abund_values = "relabundance", detection = 0.1/100, prevalence = 50/100,
```
All the arguments passed onto `getPrevalentTaxa` should be passed on by using the dots `...`

TuomasBorman replied:
All arguments that are used as arguments of getPrevalentTaxa --> ...

TuomasBorman replied:
All getPrevalentTaxa's arguments --> ...

FelixErnst commented on R/prevalentAbundance.R:
```diff
@@ -0,0 +1,76 @@
+                   include_lowest = FALSE, as_relative=FALSE, ...){
+
+              #Adds relative abundance table to the data
+              x <- relAbundanceCounts(x)
```
No point in adding it, since you don't return `x`. Use the internal function `.calc_rel_abund` to get the matrix and add a dispatch for `ANY` on `getPrevalentTaxa`. See `getPrevalence` for an example of implementation

FelixErnst commented on R/prevalentAbundance.R:
```diff
@@ -0,0 +1,76 @@
+              }
+
+              # Pick the core and calculate abundance
+              if (is.vector(values)) {
```
`values` is and always will be a `matrix`. This whole if-part doesn't make sense

TuomasBorman replied:
Deleted the if structure. Now it is only

# Pick the core and calculate abundance
              ret <- colSums(values[cm, ])

FelixErnst commented on R/prevalentAbundance.R:
```diff
@@ -0,0 +1,76 @@
+                  ret <- values # 1 OTU x N samples
+              } else if (length(cm) == 1) {
+                  ret <- values[cm,]
+              } else if (ncol(values) > 1 && length(cm) > 1) {
```
Since `values` is always a `matrix`, `colSums` will always work. Use `colSums(values[cm, ,drop=FALSE])` If you need to. 

Hint - this code is the solution to this whole code block from line 63-74:
```
colSums(values[cm, ,drop=FALSE])
```

Replace these lines with this and it will always work.

TuomasBorman replied:
Changed to 

ret <- colSums(values[cm, ,drop=FALSE])

FelixErnst commented on R/estimateDominance.R:
```diff
@@ -0,0 +1,366 @@
+              #Initialize table that is used to store indices
+              tab <- NULL
+
+              #If the indices list is not null
```
I think it is not a list, but a vector

TuomasBorman replied:
You're right --> corrected the comment.

FelixErnst reviewed (changes requested):
Hi Tuomas, Hi Leo,

thanks for the PR.

This is a nice additon, but it needs some refactoring. Please see the comments for more details and let me know if some things need to be clarified.

edit: also test for the functions, also the internal functions, are required.

Felix

TuomasBorman committed:
Synced with FelixErnst/mia

Merge remote-tracking branch 'upstream/master' into tvborm

# Conflicts:
#	R/agglomerate.R

TuomasBorman committed:
Changed the name of prevalentAbundance to getPrevalentAbundance

TuomasBorman committed:
Changed the names of internal functions. E..g checkIndices --> .check_indices and gini --> .gini

TuomasBorman committed:
Corrected a comment

TuomasBorman committed:
All arguments that are used as arguments of getPrevalentTaxa --> ...

TuomasBorman committed:
Changed roxygen comments

TuomasBorman committed:
Removed BPPARAM argument because it is not used in the code

TuomasBorman committed:
.gini_help to .calculate_gini

TuomasBorman committed:
.gini to .get_gini

TuomasBorman committed:
Synced with FelixErnst/mia

Merge remote-tracking branch 'origin/master' into tvborm

# Conflicts:
#	DESCRIPTION
#	R/estimateDominance.R
#	man/estimateDominance.Rd

TuomasBorman committed:
Deleted vector checking part

TuomasBorman committed:
Warning changed to error

TuomasBorman committed:
Changed warning to error

TuomasBorman committed:
Minor changes

TuomasBorman committed:
Synced with FelixErnst/mia

Merge remote-tracking branch 'upstream/master' into tvborm

# Conflicts:
#	DESCRIPTION

TuomasBorman committed:
Synced with FelixErnst/mia

Merge remote-tracking branch 'upstream/master' into tvborm

# Conflicts:
#	DESCRIPTION

TuomasBorman committed:
Changed roxygen comments

TuomasBorman committed:
assays() functions changed to assay()

TuomasBorman committed:
Synced with FelixErnst/mia

Merge remote-tracking branch 'upstream/master' into tvborm
# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

TuomasBorman committed:
Changed check_indices function to more simple solution

Merge remote-tracking branch 'upstream/master' into tvborm

TuomasBorman committed:
Changed functionality of gini function, corrected one typo

TuomasBorman committed:
Implemented major changes from estimateDiversity

TuomasBorman committed:
Implemented abund_values check

TuomasBorman committed:
Synced with FelixErnst/mia

Merge remote-tracking branch 'upstream/master' into tvborm
# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

TuomasBorman committed:
Added checking of arguments

TuomasBorman committed:
Deleted unused internal function

TuomasBorman committed:
Synced with FelixErnst/mia, changed rank back that it gets values 1 or 2

Merge remote-tracking branch 'upstream/master' into tvborm
# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

TuomasBorman committed:
Changed the name of 'rank' to 'ntaxa'

TuomasBorman committed:
Changed warning to error

TuomasBorman committed:
Changed the default value of index in .dominance_help to NULL, other minor changes

TuomasBorman committed:
Synced with FelixErnst/mia

FelixErnst committed:
refactored estimateDominance and related functions

FelixErnst commented:
Hi @tvborm 

I had a look at the changes. The structure is better, but since we want to be able to calculate all index from the same call it requires some additional changes.

Therefore, I changed some things around, which you should have a look at and comment, whether you think, it makes sense.

In addition, I noticed the follwing things:
-  the argument `as_relative` is always defined by a value of index (dropped for gini, simpson and prevalence and set by the other ones) Therefore, I removed it.
- the dispatch to `MicrobiomeExperiment` is unnecessary, since we dont use any sequence of tree data in the functions. Therefore, the `SummarizedExperiment` dispatch is enough for now
- the `getPrevalentTaxa` needed a tweak to work with a simple matrix. See changes there

If you are done, please add test for all the functions. I added already a file in tests.

Felix

TuomasBorman committed:
Added unit tests for estimateDominance

TuomasBorman committed:
Minor change to test

FelixErnst commented on tests/testthat/test-5dominance.R:
```diff
@@ -0,0 +1,29 @@
+context("estimateDominance")
+
+test_that("estimateDominance", {
```
Unit test should be structured slighlty differently. The main focus is to make sure it work as intended, not that it works.

Two way exists:
- Either you compute it by two different ways
- you compare the results to fixed value, you know to be true. (Use pen and paper if you have too)

The examples below show that it works but not necessarily as intended. In the follwoing code chunk is an example, which you can use as blue print for adding all the tests.
```
# .simpson_dominance
x <- 1
expect_equal(mia:::.simpson_dominance(x),1)
x <- 1:1000
expect_equal(mia:::.simpson_dominance(x), 0.001332667)
```

So in this style you need to make sure that the follwing functions work correctly:
- `.gini_dominance`
- `.get_core_dominance`
- `.get_dominance` for all four cases

then you check if all indeces get calculated for the call `estimateDominance(esophagus)` by checking the colnames of `colData()` after making sure, that the return value of `estimateDominance` is actually the same type as the input.

edit: also don't use the slots directly. If, for whatever reason the slots change, our code will still work since it uses `colData()<-` to write the data. The tests will not since the slots get accessed directly. You can use `colData(estimateDominance(esophagus))` if you want

TuomasBorman replied:
Ok, thanks for help. I changed tests little bit. All, but 3 are working. It seems that .simpson_dominance should return value 1-a,  not a.

FelixErnst replied:
And I really like that you test against vegan functions. You can fix the error in the code of `.simpson_dominance`, if you are sure about it. @antagomir

antagomir replied:
Simpson is confusing bc there are [different indices](https://en.wikipedia.org/wiki/Diversity_index#Simpson_index).

The _Simpson index_ is defined as $\lambda = \sum_i p^2_i$. 

_Simpson dominance index_ is defined as $\lambda$. It is the same as Simpson index (one source [here](https://www.sciencedirect.com/science/article/pii/S2405844019362668)

`Inverse Simpson` diversity index is defined as $\frac{1}{\lambda}$.

Then we also have `Gini-Simpson` index $1 - \lambda$.

Would be nice to have all of these in the package but the definitions must be clear in order to avoid misuse. Just check that the function corresponds to these official definitions (inverse Simpson and Gini-Simpson should go to `estimateDiversity` whereas Simpson should go to `estimateDominance`).

Another point to check that all dominance indices (and all diversity indices) have the same sign. In some cases the interpretation of alpha diversity/dominance indices can be inverse compared to the other indices. Where this is the case, we may like to consider how to treat. It is causing some risks for interpretation if the indices with the same purpose have opposite interpretations for increasing index values (higher values meaning either decreasing or increasing diversity or dominance). Could you check on the same go if there are such cases so that we can see what to do. That discussion can go in a new issue.

TuomasBorman replied:
Modified simpson tests. I calculated simpson index with pen and paper and tested .simpson_dominance with that

TuomasBorman committed:
Changed unit tests

FelixErnst commented on tests/testthat/test-5dominance.R:
```diff
@@ -2,28 +2,72 @@ context("estimateDominance")
+    #Checks that the type of output is the same as the type of input.
+    expect_true(typeof(tse_idx) == typeof(tse))
+    #Checks that every index is calculated by checking the column names from colData.
+    expect_true(all(
```
the function `expect_named` is available as well. `expect_named(colData(tse_idx),c("relative", ....... ))`. Be aware, that correct order is tested as well.

TuomasBorman replied:
I changed it to 

expect_named(colData(tse_idx), c("DBP", "DMN", "absolute", "relative", "simpson", "core_abundance", "gini"))

estimateDominance calculates the indices in the same order as they are in the "Index" vector. I assumed that it is the right order.

antagomir committed:
sync

antagomir committed:
sync

antagomir committed:
reverting the changes and syncing

TuomasBorman committed:
Modified column/index name test

FelixErnst commented:
I think the last open topic is the correct value for the simpson domince. Can you comment/check on this?

The PR is ready to be merged.

FelixErnst commented:
One additional thing, the package fails the test. You have to include the `reldist` package in the suggest and add `skip_if_not_installed` call at the beginning of the dominance test, which rely on the comparison to the `reldist` package

TuomasBorman committed:
Modified simpson tests

TuomasBorman committed:
Added stuff related to reldist package requirement

FelixErnst committed:
version bump

FelixErnst commented:
I bumped the version, in order to force an dependency update in GHA.

FelixErnst committed:
added Tuomas Borman as ctb
